### PR TITLE
Add usage limits fields for new discount system

### LINKED ISSUE #38657: Set a usage limit per customer
# Context
The user can set a usage limit per customer (integer >= 1).
By default, this limit is 1.
An option allows for unlimited usage per customer.
If a discount is intended to a single customer, this option is not displayed.
Once the limit of use/application is reached (https://github.com/PrestaShop/PrestaShop/issues/38663), customers can no longer benefit from the discount, even if the limit of usage for those customers is not reached.

# Mockup

Image

# Tests
## Field validation
**GIVEN:** an invalid usage limit 
**WHEN:** I want to create the discount
**THEN:** the discount is not created and a error message is displayed

**GIVEN:** a valid usage limit
**WHEN:** I want to create the discount
**THEN:** the discount is created

**GIVEN:** a limit of use set to 50
**WHEN:** I set a usage limit to 49
**THEN:** the customer limit is valid

**GIVEN:** a limit of use set to 50
**WHEN:** I set a usage limit to 51
**THEN:** the customer limit is not valid

**GIVEN:** a discount intended to a single customer
**WHEN:** I want to set a usage limit
**THEN:** the field is not diplayed

## Apply discount
**GIVEN:** the usage limit for a customer is reached
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the usage limit for a customer is reached 
**THEN:** the discount is not applied

**GIVEN:** the usage limit for a customer is not reached yet
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is applied

**GIVEN:** a discount applied automatically
**WHEN:** the usage limit for a customer is not reached yet
**THEN:** the discount is applied

**GIVEN:** the usage limit for a customer is not reached yet, but the limit of use/application is reached
**WHEN:** I want to apply the discount via promo code (with the same customer)
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the usage limit for a customer is not reached yet, but the limit of use/application is reached
**THEN:** the discount is not applied

### LINKED ISSUE #38663: Set a limit on the application/use of discounts
# Context
Merchants can specify whether a discount can be applied/used without limitation or set a limit.
This limit must be an integer greater than or equal to 1 (1 by default).

If the discount is intended for all customer or specific customer group(s), only the first X (= limit) customers could benefit from the discount.
If the discount is intended for a single customer, only the same customer could benefit X (=limit) time(s) from the discount.

So here we in the end have 3 cases :

    limit total per discount => 100 uses that can all be used by any user ( one or more )
    limit per user => no uses limit on the discount but each user can use it x amount of time
    the combination of both => common pile => I have 100 discount available each user can use it 20 times => if there is only 15 discount left the last user can only use it 15 times.

Design wise, both field should be placed next to each other and the wording of the per customer ( 1st case ) should be something like "total uses of discounts"

# Mockup
TBUP

# Tests
## Field validation
**GIVEN:** an invalid limit 
**WHEN:** I want to create the discount
**THEN:** the discount is not created and a error message is displayed

**GIVEN:** a valid limit
**WHEN:** I want to create the discount
**THEN:** the discount is created

## Apply discount
**GIVEN:** the limit is reached
**WHEN:** I want to apply the discount via promo code
**THEN:** the discount is not applied and a error message is displayed instead

**GIVEN:** a discount automatically applied
**WHEN:** the limit is reached 
**THEN:** the discount is not applied

**GIVEN:** the limit is not reached yet
**WHEN:** I want to apply the discount via promo code
**THEN:** the discount is applied

**GIVEN:** a discount applied automatically
**WHEN:** the limit is not reached yet
**THEN:** the discount is applied

### LINKED ISSUE #40284: Customer usage limit
### Description

Add support for customer usage limits in the Discount API. Customer usage limits allow merchants to restrict how many times a specific customer can use a discount.

**Required Changes:**
- Add `available_per_user` field to the API resource schema
- Support reading usage limits via GET endpoints
- Support writing usage limits via POST/PUT endpoints
- Ensure proper validation (must be positive integer or 0 for unlimited)

**Technical Details:**
- Field: `cart_rule.quantity_per_user`
- Value 0 means unlimited usage per customer
- Value > 0 means the customer can use the discount that many times
- Usage is tracked per customer via orders and cart applications
- The system checks usage when applying discounts to carts

#### Prerequisites
1. Set up a PrestaShop 9 instance with API access configured
2. Create an API client with appropriate permissions
3. Create test customers:
   - Customer A (ID: 1)
   - Customer B (ID: 2)
4. Set up a test product and cart functionality

#### Test Case 1: Create Discount with Usage Limit
1. **Create a discount with usage limit:**
   ```bash
   POST /api/discounts
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "name": "Limited Use Discount",
     "code": "LIMIT3",
     "active": true,
     "conditions": {
       "available_per_user": 3
     },
     "discount": {
       "reduction": {
         "type": "percentage",
         "value": 10
       }
     }
   }
   ```

2. **Verify creation:**
   - Response status should be 201 Created
   - Response should include `conditions.available_per_user: 3`
   - Verify in Back Office that the usage limit is set correctly

#### Test Case 2: Read Usage Limit via GET
1. **Retrieve a discount with usage limit:**
   ```bash
   GET /api/discounts/{discountId}
   Authorization: Bearer {api_token}
   ```

2. **Verify response:**
   - Response should include `conditions.available_per_user` field
   - Value should match what was set during creation

#### Test Case 3: Update Usage Limit
1. **Update an existing discount's usage limit:**
   ```bash
   PUT /api/discounts/{discountId}
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "conditions": {
       "available_per_user": 5
     }
   }
   ```

2. **Verify update:**
   - Response status should be 200 OK
   - Response should reflect the updated usage limit
   - Verify in Back Office that the limit was updated

#### Test Case 4: Set Unlimited Usage (0)
1. **Create or update discount with unlimited usage:**
   ```bash
   PUT /api/discounts/{discountId}
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "conditions": {
       "available_per_user": 0
     }
   }
   ```

2. **Verify unlimited usage:**
   - Response status should be 200 OK
   - Response should show `available_per_user: 0`
   - Customer should be able to use the discount unlimited times

#### Test Case 5: Validation - Negative Value
1. **Attempt to set negative usage limit:**
   ```bash
   POST /api/discounts
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "conditions": {
       "available_per_user": -1
     },
     ...
   }
   ```

2. **Verify validation:**
   - Response status should be 400 Bad Request
   - Response should include validation error about invalid value

#### Test Case 6: Validation - Non-Integer Value
1. **Attempt to set non-integer usage limit:**
   ```bash
   POST /api/discounts
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "conditions": {
       "available_per_user": 3.5
     },
     ...
   }
   ```

2. **Verify validation:**
   - Response status should be 400 Bad Request
   - Response should include validation error about integer requirement

#### Test Case 7: Functional Test - Usage Limit Enforcement
1. **Create a discount with usage limit of 2:**
   ```bash
   POST /api/discounts
   Authorization: Bearer {api_token}
   Content-Type: application/json
   
   {
     "name": "Two-Time Use Discount",
     "code": "USE2X",
     "active": true,
     "conditions": {
       "available_per_user": 2
     },
     "discount": {
       "reduction": {
         "type": "percentage",
         "value": 15
       }
     }
   }
   ```

2. **Test usage tracking:**
   - As customer 1, use the discount in cart/order #1: Should succeed
   - As customer 1, use the discount in cart/order #2: Should succeed
   - As customer 1, attempt to use the discount in cart/order #3: Should fail with "usage limit reached" error
   - As customer 2, use the discount: Should succeed (separate limit per customer)

#### Test Case 8: Check Remaining Usage
1. **After creating a discount with usage limit, check remaining usage:**
   ```bash
   GET /api/discounts/{discountId}/usage?customer_id=1
   Authorization: Bearer {api_token}
   ```

2. **Verify response:**
   - Response should include:
     - `total_limit`: 3 (or whatever was set)
     - `used_count`: 1 (or current usage)
     - `remaining`: 2 (or remaining uses)

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add usage limits fields for new discount system.<br><br>**Bonus:** add the ability to add a custom label for disabling_switch component.<br/>**Bonus2:** introduce a DirtyTrait to track modified properties in an object.<br>**Bonus3:** add tests for usage limits in FO (and validate the `null` state too for no limit!)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See #38663 & #38657
| UI Tests          | https://github.com/boherm/ga.tests.ui.pr/actions/runs/20229798509
| Fixed issue or discussion?     | Fixes #38663 Fixes #38657
| Related PRs       | https://github.com/PrestaShop/ps_apiresources/pull/119
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Catalog.Feature`
    - [ ] `Specifiy discount mode`
    - [ ] `Select total usage limits`
    - [ ] `No limit`
    - [ ] `Select usage limits per customer`
- `Admin.Catalog.Help`
    - [ ] `This discount can be used "X" time(s) in total.`
    - [ ] `A customer will only be able to use this discount "X" time(s).`
- `Admin.Catalog.Notification`
    - [ ] `The usage limit per customer must be less than or equal to the total usage limit.`

(Note: this is an automated message, but answering it will reach a real human)

File: classes/CartRule.php [hunk 1]
```diff
@@ -1982,7 +1982,7 @@ public static function autoAddToCart(?Context $context = null, bool $useOrderPri
 		LEFT JOIN ' . _DB_PREFIX_ . 'cart_rule_country crco ON cr.id_cart_rule = crco.id_cart_rule
 		WHERE cr.active = 1
 		AND cr.code = ""
-		AND cr.quantity > 0
+		AND (cr.quantity > 0 OR cr.quantity is null)
```
Comment (jolelievre) [inline classes/CartRule.php:1979]: Did you search in other files (especially legacy ones) if these two fields are used? especially in conditions?

File: src/Adapter/Discount/Update/Filler/DiscountFiller.php [hunk 2]
```diff
@@ -38,59 +38,59 @@ class DiscountFiller
     public function fillUpdatableProperties(CartRule $cartRule, UpdateDiscountCommand $command): array
     {
         $updatableProperties = [];
-        if (null !== $command->getValidFrom()) {
+        if ($command->isDirty('validFrom')) {
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/Filler/DiscountFiller.php:41]: I'm not sure this is required for all use cases 🤔
Technically, it's only meaningful for nullable properties, no?

Not that it impacts in a bad way anyway.

File: src/Core/Domain/Discount/Command/UpdateDiscountCommand.php [hunk 3]
```diff
@@ -94,6 +98,7 @@ public function isActive(): ?bool
     public function setActive(bool $active): self
     {
         $this->active = $active;
+        $this->markDirty('active');
```
Comment (jolelievre) [inline src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:101]: Same question as above, do we need to handle the `markDirty` for all fields? Or only the ones that are relevant and have a meaningful null value?

File: src/Core/Domain/Discount/Command/UpdateDiscountCommand.php [hunk 4]
```diff
@@ -133,6 +140,8 @@ public function setValidityDateRange(DateTimeImmutable $from, DateTimeImmutable
 
         $this->validFrom = $from;
         $this->validTo = $to;
+        $this->markDirty('validFrom');
+        $this->markDirty('validTo');
```
Comment (jolelievre) [inline src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:144]: I guess this will be useful for `validTo` that should also be turned into a nullable in another PR

File: src/Core/Trait/DirtyTrait.php [hunk 5]
```diff
@@ -24,45 +24,51 @@
...
     /**
-     * @param array<int, int> $tagIds
+     * Marks a specific property as dirty (modified).
+     *
+     * @param string $propertyName
+     *
+     * @return void
      */
-    public function __construct(array $tagIds)
+    public function markDirty(string $propertyName): void
```
Comment (jolelievre) [inline src/Core/Trait/DirtyTrait.php:48]: Out of curiosity, you didn't find a way to automatically mark properties as dirty with `__call` for example? But I guess it's only called when the method doesn't exist right?

File: src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php [hunk 6]
```diff
@@ -44,6 +44,59 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                    'class' => 'd-flex align-items-center',
+                ],
+                'disabling_switch' => true,
+                'disabling_switch_label' => $this->trans('No limit', 'Admin.Catalog.Feature'),
+                'switch_state_on_disable' => 'on',
+                'default_empty_data' => null,
+                'disabled_value' => function ($data) {
+                    return null === $data;
+                },
+                'disabling_switch_event' => 'quantityTotalSwitchNolimit',
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php]: This event doesn't seem to be used in JS, maybe it's overkill? (same for the second event)

File: src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php [hunk 7]
```diff
@@ -44,6 +44,59 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                'default_empty_data' => null,
+                'disabled_value' => function ($data) {
+                    return null === $data;
+                },
+                'disabling_switch_event' => 'quantityPerCustumerSwitchNolimit',
+                'constraints' => [
+                    new Assert\GreaterThanOrEqual(0),
+                    new Assert\LessThanOrEqual([
+                        'propertyPath' => 'parent.all[quantity_total].data',
+                        'message' => $this->trans('The usage limit per customer must be less than or equal to the total usage limit.', 'Admin.Catalog.Notification'),
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php]: Suggestion for wording, maybe easier to read this way
```suggestion
                        'message' => $this->trans('The usage limit per customer cannot be higher than the total usage limit.', 'Admin.Catalog.Notification'),
```

File: tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php [hunk 8]
```diff
@@ -92,6 +92,14 @@ public function assertStartDateCannotBeAfterExpiryDate(): void
         $this->assertLastErrorIs(DiscountConstraintException::class, DiscountConstraintException::DATE_FROM_GREATER_THAN_DATE_TO);
     }
 
+    /**
+     * @Then I should get an error that the customer can't use the discount anymore
+     */
+    public function assertDiscountCannotBeInAddedInCart(): void
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php]: There is already a step for that `I should get an error that the discount is invalid`

File: src/Adapter/Discount/Update/Filler/DiscountFiller.php (hunk 2 above)
Comment (boherm) [reply src/Adapter/Discount/Update/Filler/DiscountFiller.php:41]: To me, if we use this new trait, it's better to keep all dirty properties and not only one 🤔

File: src/Core/Domain/Discount/Command/UpdateDiscountCommand.php (hunk 3 above)
Comment (boherm) [reply src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:101]: Same answer as above :D

File: src/Core/Trait/DirtyTrait.php (hunk 5 above)
Comment (boherm) [reply src/Core/Trait/DirtyTrait.php:48]: Yep, but it doesn't work. `__call` or `__set` seems to be called only if the function doesn't exist in the class.
So, we need to have here a `markDirty` function to track theses properties.

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Catalog.Feature`
    - [ ] `Specifiy discount mode`
    - [ ] `Select total usage limits`
    - [ ] `No limit`
    - [ ] `Select usage limits per customer`
- `Admin.Catalog.Help`
    - [ ] `This discount can be used "X" time(s) in total.`
    - [ ] `A customer will only be able to use this discount "X" time(s).`
- `Admin.Catalog.Notification`
    - [ ] `The usage limit per customer cannot be higher than the total usage limit.`

(Note: this is an automated message, but answering it will reach a real human)

File: src/Core/Domain/Discount/Command/UpdateDiscountCommand.php (hunk 3 above)
Comment (jolelievre) [reply src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:101]: Alright!

**Review by jolelievre — DISMISSED** [review]
Thanks

**Review by tleon — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]

**Review by tleon — DISMISSED** [review]

**Review by nicosomb — DISMISSED** [review]
